keps/sig-api-machinery/555-server-side-apply/README.md

```diff
@@ -279,6 +286,172 @@ The conversion between the two and creating the diff was complex and would have
+you need any help or guidance.
+
+-->
+
+### Feature Enablement and Rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: [ServerSideApply](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/features/kube_features.go#L100)
+    - Components depending on the feature gate: kube-apiserver
+
+* **Does enabling the feature change any default behavior?**
+
+ While this changes how objects are modified and then stored in the database, all the changes should be strictly backward compatible, and shouldn’t break existing automation or users. The increase in size can possibly have adverse, surprising consequences including increased memory usage for controllers, increased bandwidth usage when fetching objects, bigger objects when displaying for users (kubectl get -o yaml). We’re trying to mitigate all of these with the addition of a new header.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+  Describe the consequences on existing workloads (e.g., if this is a runtime
+  feature, can it break the existing applications?).
+
+  Yes. Managed fields will be reset for server-side applied objects.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+
+The feature will be restored. Server-side applied objects will have lost their “set” which may cause some surprising behavior (fields might not be removed as expected).
+
+* **Are there any tests for feature enablement/disablement?**
+  The e2e framework does not currently support enabling or disabling feature
+  gates. However, unit tests in each component dealing with managing data, created
+  with and without the feature, are necessary. At the very least, think about
+  conversion tests if API types are being modified.
+
+  Tests are in place for upgrading from client side to server side apply and vice versa.
+
+### Rollout, Upgrade and Rollback Planning
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  Try to be as paranoid as possible - e.g., what if some components will restart
+   mid-rollout?
+There is no specific way that the rollout can fail. The rollout can't impact existing workload.
+* **What specific metrics should inform a rollback?**
+
+The feature shouldn't affect any existing behavior. A surprisingly high number of modification rejections could be a sign that something is not working properly.
+
+* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
+Because the feature doesn't affect existing behavior, rollback and upgrades haven't be specifically tested.
```

Comment:
One of the requirements for GA is converting at least one controller.  At the moment, I don't know of any "only available in GA" feature of SSA it would use, but if one is used, this becomes a problem right?

Mind listing the risk (not the problem).

Reply:
Good point! Let me think about it.

Reply:
OK, @jpbetz @Jefftree and I looked at it, we believe it's fine!

```suggestion
Because the feature doesn't affect existing behavior, rollback and upgrades haven't be specifically tested. 
The feature is being used by the cluster role aggregator though. Upgrading/downgrading/upgrading, which
could result in the managedFields being removed, wouldn't cause any problem since the `Rules` fields
filled by the controller is `atomic`, and thus doesn't depend on the current state of the managedFields.
```